File: Minimum Jumps.java
class Solution {
    static int minJumps(int[] arr) {
        // your code here
        
        int jumps=0;
        int currentIndex=0;
        int farIndex=0;
        
        //if arr length is 0 or 1st element is 0
        if(arr.length==0 || arr[0]==0) return -1;
        
        //for length 1
        else if(arr.length==1) return 1;
        
        else {
            for(int i=0; i<arr.length-1; i++){
                farIndex=Math.max(farIndex, i+arr[i]);
                
                if(i==currentIndex){
                    jumps++;
                    currentIndex=farIndex;
                    
                    if(currentIndex>=arr.length-1) return jumps;
                    
                }
                
            }
        }
        return -1;
    }
}

File: Minimum Jumps.cpp
class Solution {
  public:
    int fun(int i, vector<int> &a, vector<int> &dp){
        if(i==a.size()-1)
            return 0;
        if(dp[i]!=-1)
            return dp[i];
        
        int res = INT_MAX;
        for(int j = i+1 ; j<=min(i+a[i], int(a.size()-1));j++){
            int c = fun(j, a, dp);
            
            if(c!=INT_MAX)
                res = min(res,1+c);
        }
        
        return dp[i] = res;
    }
  
    int minJumps(vector<int>& arr) {
        // code here
        int n = arr.size();
        vector<int> dp(n+1, -1);
        
        int ans = fun(0, arr, dp);
        return ans==INT_MAX?-1:ans;
    }
};

